src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs

```diff
@@ -0,0 +1,143 @@
+    /// job object.
+    /// </summary>
+    /// <param name="process">The process to add to the job.</param>
+    /// <returns>Whether the job creation and assignment worked or not.</returns>
+    public bool AssignProcessToJobObject(SafeProcessHandle process)
+        => InitializeJob() && Interop.Windows.AssignProcessToJobObject(_jobObject, process);
+
+    /// <summary>
+    /// Blocks the current thread until all processes in the job have exited.
+    /// </summary>
+    /// <param name="cancellationToken">A token to cancel the operation.</param>
+    public void WaitForExit(CancellationToken cancellationToken)
+    {
+        if (_completionPort is null)
+        {
+            return;
+        }
+
+        using var cancellationRegistration = cancellationToken.Register(() =>
+        {
+            Interop.Windows.PostQueuedCompletionStatus(
+                _completionPort,
+                Interop.Windows.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO,
+                lpCompletionKey: nint.Zero,
+                lpOverlapped: nint.Zero);
+        });
+
+        const int INFINITE = -1;
+        int completionCode = 0;
+        do
+        {
+            Interop.Windows.GetQueuedCompletionStatus(
+                _completionPort,
+                out completionCode,
+                lpCompletionKey: out _,
+                lpOverlapped: out _,
+                INFINITE);
+        }
+        while (completionCode != Interop.Windows.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO);
+        cancellationToken.ThrowIfCancellationRequested();
+    }
+
+    [MemberNotNullWhen(true, [nameof(_jobObject), nameof(_completionPort)])]
+    private bool InitializeJob()
+    {
+        if (_initStatus.HasValue)
+        {
+            return _initStatus.Value;
+        }
+
+        if (_jobObject is null)
+        {
+            _jobObject = Interop.Windows.CreateJobObject(nint.Zero, nint.Zero);
+            if (_jobObject.IsInvalid)
+            {
+                _initStatus = false;
+                return false;
+            }
+        }
+
+        if (_completionPort is null)
+        {
+            _completionPort = Interop.Windows.CreateIoCompletionPort(
+                FileHandle: -1,
+                ExistingCompletionPort: nint.Zero,
+                CompletionKey: nint.Zero,
+                NumberOfConcurrentThreads: 1);
+
+            if (_completionPort.IsInvalid)
+            {
+                _initStatus = false;
+                return false;
+            }
+        }
+
+        var completionPort = new Interop.Windows.JOBOBJECT_ASSOCIATE_COMPLETION_PORT()
+        {
+            CompletionKey = _jobObject.DangerousGetHandle(),
+            CompletionPort = _completionPort.DangerousGetHandle(),
+        };
+
+        _initStatus = Interop.Windows.SetInformationJobObjectCompletionPort(
+            _jobObject,
+            ref completionPort);
+
+        return _initStatus.Value;
+    }
+
+    ~JobProcessCollection()
+    {
+        Dispose();
+    }
```

Comment:
Extra method. We can remove since all unmanaged resources wrapped in SafeHandle.
https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#finalizable-types